src/mapping/query/functions.rs

```diff
@@ -210,6 +211,171 @@ impl Function for ToTimestampFn {
 
 //------------------------------------------------------------------------------
 
+#[derive(Debug)]
+pub(in crate::mapping) struct UpcaseFn {
+    query: Box<dyn Function>,
+}
+
+impl UpcaseFn {
+    pub(in crate::mapping) fn new(query: Box<dyn Function>) -> Self {
+        Self { query }
+    }
+}
+
+impl Function for UpcaseFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        match self.query.execute(ctx)? {
+            Value::Bytes(bytes) => {
+                let mut buf = BytesMut::with_capacity(bytes.len());
+
+                buf.extend_from_slice(&bytes);
+                buf.iter_mut().for_each(|c| c.make_ascii_uppercase());
+
+                Ok(Value::Bytes(buf.freeze()))
+            }
+            _ => Err("unable to upcase non-string types".to_string()),
```

Comment:
I think it would be worth creating a custom error type for these errors. The errors could be based on error codes rather than strings so we can locate the source of an error even if the error message text changes over versions. 

Also if the ErrorCode -> Message String mapping for the errors is kept in a single file it will be easier to keep them consistent.

Reply:
Related #978

Reply:
This may be a good place to get started on that change as the number of error messages in the system will be increasing a huge amount with this transform.

Reply:
I've renamed the error messages. I recommend holding off on changing this to use structured errors for now. I'd like to get this PR in first, and then iterate on these non-blocking suggested improvements separately.